dhis2/dhis2-core
# feat: apps can register additional protected namespaces [DHIS2-17546]

### Summary
Adds the possibility for apps to declare additional datastore namespaces that should be associated with the app. For each the user can specify the authorities required to read/write. 

In the `manifest.webapp` a new `additionalNamespaces` member is added to the same parent `activities.dhis` where an app also would declare the main `namespace` used. Each of the additional namespace objects must declare a `namespace` name, and authorities by one or more of the following properties:

* `authorities`: a user needs any of these to read or write
* `readAuthorities`: a user needs any of these to read
* `writeAuthorities`: a user needs any of these to write

If `authorities` is combined with read or write limited list the entries in `authorities` are added to the read/write (union)
If only `readAuthorities` are defined these automatically apply for write. 
If only `writeAuthorities` are defined read is not restricted. 

Examples of valid additional namespace objects
```json
"additionalNamespaces": [
  { "namespace": "extra1", "authotities": ["M_extra1"]},
  { "namespace": "extra2", "readAuthorities": ["M_extra2read"]},
  { "namespace": "extra3", "writeAuthotities": ["M_extra3write"]},
  { "namespace": "extra4", "authorities": ["M_extra4readwrite"], "writeAuthotities": ["M_extra4write"]},
]
```

### Datastore
To allow to specify the read and write authorities of a datastore namespace protection individually the existing feature got slightly extended to support passing and evaluating two individual sets of authorities instead of a common one.
This feature is however only used by the added app additional namespaces. All existing uses of namespace protection simply use the same set of authorities for read and write behaving like before when it was a common set.

### Collection Utils Move
In order to be able to use the `union` method in methods in `api-web` the utility needed to be moved to that module (or a common dependency module which we do not have). While I think the class is better located in `api-web` this unfortunately meant that lots of imports did change where the util class was used.

### Automatic Testing
The newly added `Set` `union` helper method got covered with new unit tests.

### Manual Testing
* download an app zip from the app hub (ideally something simple, I used the cache cleaner app)
* open the zip file and edit the `manifest.webapp` inside (some tools do support "in-zip" editing)
* add `additionalNamespaces`, like in the below extract example
```json
  "activities": {
    "dhis": {
      "href": "*",
      "additionalNamespaces": [
        {"namespace": "y-control", "authorities": ["Y"]},
        {"namespace": "test2", "writeAuthorities": ["M_dhis-web-maintenance"]}
      ]
    }
  },
```
* manually install the changed app
* use the namespace and verify it now requires the specified authorities to read/write
* change the file again and add/change so that a declaration is incomplete, e.g. no/empty `namespace` or no/empty `authorities`
* reinstall the changed app and verify the install fails with an error pointing to the cause
* download another app and update its `manifest.webapp` file to use the same namespace that is already used by the first app
* try to manually install that app and verify it fails due to the namespace already being associated with another app
